### PR TITLE
[Feature] Edge Feature Support on gsf.py and embed.py

### PR DESCRIPTION
**Pull request by zhjwy9343**
*Issue #, if available:*

*Description of changes:*
This PR is the 2nd PR for supporting edge features. This PR,

add an util function get_edge_feat_size in gsf.py;
add edge input layer in embed.py;
add get edge feature for blocks in dataloading.py;
add intialization of edge input layer in set_encoder() in gsf.py.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/model/embed.py [hunk 1]
```diff
@@ -469,6 +539,148 @@ def use_wholegraph_sparse_emb(self):
...
+                    f"We need a projection weight for edge features of edge type {canonical_etype}"
+                emb = feats.float() @ self.input_projs[str(canonical_etype)]
+
+                if self.activation is not None:
+                    emb = self.activation(emb)
+                emb = self.dropout(emb)
+                embs[canonical_etype] = emb
+
+            if self.num_ffn_layers_in_input > 0:
+                embs = {can_etype: self.ngnn_mlp[str(can_etype)](h) \
```
Comment (classicsong) [inline python/graphstorm/model/embed.py:666]:
```suggestion
                embs = {canonical_etype: self.ngnn_mlp[str(canonical_etype)](h) \
```

File: tests/unit-tests/test_dataloading.py [hunk 2]
```diff
@@ -453,6 +453,88 @@ def test_GSgnnData2():
...
+            ('n0', 'r0', 'n1'): ['none'],
+            ('n0', 'r1', 'n1'): ['none']
+        }
+        for _, _, blocks in dataloader:
+            try:
+                edge_feat_list = gdata.get_blocks_edge_feats(blocks, efeat_fields)
+            except:
+                edge_feat_list = {}
+            assert len(edge_feat_list) == 0
+
```
Comment (classicsong) [inline tests/unit-tests/test_dataloading.py:534]: Can we also test the case when there are two edge features and both exists in the graph? And also the case when layer 0 does not have edge feature but layer 1 does?

File: tests/unit-tests/test_embed.py [hunk 3]
```diff
@@ -176,6 +179,97 @@ def test_input_layer3(dev):
...
+    block_edge_feat_list.append(feat1)
+    feat2 = {
+        ('n0', 'r0', 'n1'): g.edges[('n0', 'r0', 'n1')].data['feat'][np.arange(10, 14)].to(dev),
+        ('n0', 'r1', 'n1'): g.edges[('n0', 'r1', 'n1')].data['feat'][np.arange(10, 14)].to(dev)
+    }
+    block_edge_feat_list.append(feat2)
+    nn.init.eye_(edge_input_layer.input_projs[str(('n0', 'r0', 'n1'))])
+    nn.init.eye_(edge_input_layer.input_projs[str(('n0', 'r1', 'n1'))])
+    embed = edge_input_layer(block_edge_feat_list)
+
```
Comment (classicsong) [inline tests/unit-tests/test_embed.py:240]: Can you also test the case when 1st layer is None and the 2nd layer has edge features?

File: python/graphstorm/gsf.py [hunk 4]
```diff
@@ -472,10 +541,13 @@ def create_builtin_node_model(g, config, train_task):
     -------
     GSgnnModel : The GNN model.
     """
-    if config.training_method["name"] == "glem":
+    training_method = config.training_method["name"]
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/gsf.py:544]: We will get a new lint error of `possibly-used-before-assignment` if unchanged. Although our current pylint version 2.7.15 does not raise this error, but later if upgrade pylint, we will get this error.

File: python/graphstorm/gsf.py [hunk 5]
```diff
@@ -873,41 +945,51 @@ def set_encoder(model, g, config, train_task):
...
+                                            activation=config.input_activate,
+                                            use_node_embeddings=config.use_node_embeddings,
+                                            force_no_embeddings=config.construct_feat_ntype,
+                                            num_ffn_layers_in_input=config.num_ffn_layers_in_input,
+                                            use_wholegraph_sparse_emb=config.use_wholegraph_embed)
+        # set edge encoder input layer no matter if having edge feature names or not
+        # TODO: add support of languange models and GLEM
+        edge_feat_size = get_edge_feat_size(g, config.edge_feat_name)
+        edge_encoder = GSEdgeEncoderInputLayer(g, edge_feat_size, config.hidden_size,
+                                        dropout=config.dropout,
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/gsf.py:984]: If indent, will trigger a `line-too-long` lint error.

File: python/graphstorm/gsf.py [hunk 6]
```diff
@@ -16,6 +16,7 @@
     GSF utility functions.
 """
 
+
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/gsf.py:19]: removed.

File: python/graphstorm/model/embed.py [hunk 7]
```diff
@@ -469,6 +539,148 @@ def use_wholegraph_sparse_emb(self):
         return self._use_wholegraph_sparse_emb
 
 
+class GSEdgeEncoderInputLayer(GSEdgeInputLayer):
+    """ The edge encoder input layer for edges with features in a heterogeneous graph.
+
+    The input layer adds a linear layer on edges with edge features and the linear layer
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/embed.py:545]: modified to `project layer`.

File: python/graphstorm/model/embed.py [hunk 8]
```diff
@@ -469,6 +539,148 @@ def use_wholegraph_sparse_emb(self):
...
+        # ngnn
+        self.num_ffn_layers_in_input = num_ffn_layers_in_input
+        self.ngnn_mlp = nn.ModuleDict({})
+
+        # set projection weights on edge features
+        for canonical_etype in g.canonical_etypes:
+            feat_dim = 0
+            if self.feat_size[canonical_etype] > 1:
+                feat_dim = self.feat_size[canonical_etype]
+            if feat_dim > 0:
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/embed.py:624]: removed this line and `feat_dim = 0` line.

File: python/graphstorm/model/embed.py [hunk 9]
```diff
@@ -469,6 +539,148 @@ def use_wholegraph_sparse_emb(self):
...
+                input_projs = nn.Parameter(th.Tensor(feat_dim, self.embed_size))
+                nn.init.xavier_uniform_(input_projs, gain=nn.init.calculate_gain("relu"))
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/embed.py:626]: removed.

File: python/graphstorm/model/embed.py [hunk 10]
```diff
@@ -469,6 +539,148 @@ def use_wholegraph_sparse_emb(self):
...
+            The projected edge embeddings in the format of
+            [{can_etype1: embs, can_etype2: embs, ...}, {can_etype1: embs, ...}, ...].
+        """
+        assert isinstance(block_edge_input_feats, list), \
+            'The edge input features should be in a list.'
+
+        embs_list = []
+        for edge_input_feats in block_edge_input_feats:
+            assert isinstance(edge_input_feats, dict), \
+                'The input features should be in a dict.'
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/embed.py:653]: added.

File: python/graphstorm/model/embed.py [hunk 11]
```diff
@@ -469,6 +539,148 @@ def use_wholegraph_sparse_emb(self):
...
+        embs_list: list of dicts
+            The projected edge embeddings in the format of
+            [{can_etype1: embs, can_etype2: embs, ...}, {can_etype1: embs, ...}, ...].
+        """
+        assert isinstance(block_edge_input_feats, list), \
+            'The edge input features should be in a list.'
+
+        embs_list = []
+        for edge_input_feats in block_edge_input_feats:
+            assert isinstance(edge_input_feats, dict), \
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/embed.py:652]: yes, it will be an empty dict `{}`, which means no edge feature for a specific layer.

File: tests/unit-tests/test_dataloading.py (hunk 2 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_dataloading.py:534]: added two new test. Test case 5 in `test_GSgnnData_edge_feat()`, and Test case 1 in `test_GSgnnData_edge_feat2()`

File: python/graphstorm/model/embed.py [hunk 12]
```diff
@@ -163,6 +164,75 @@ def use_wholegraph_sparse_emb(self):
         return False
 
 
+class GSEdgeInputLayer(GSLayer):  # pylint: disable=abstract-method
+    """ The base input layer for edges in a heterogeneous graph.
+
+    This is the base class for all edge input layer classes. This implementation only
+    support edge feature transformation. Unlike ``GSNodeInputLayer``, for featureless edges,
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/embed.py:171]: updated.

File: python/graphstorm/dataloading/dataset.py [hunk 13]
```diff
@@ -416,6 +416,39 @@ def get_edge_feats(self, input_edges, efeat_fields, device='cpu'):
...
+        """
+        block_edge_input_feats = []
+        for block in input_blocks:
+            input_edges = {}
+            for etype in block.canonical_etypes:
+                eid = block.edges[etype].data[dgl.EID]
+                input_edges[etype] = eid
+            edge_feat = self.get_edge_feats(input_edges, efeat_fields, device)
+            if edge_feat:
+                block_edge_input_feats.append(edge_feat)
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/dataloading/dataset.py:448]: modified to return empty dict for none edge_feat.

File: tests/unit-tests/test_embed.py (hunk 3 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_embed.py:240]: added one more, Test case 4.

File: python/graphstorm/gsf.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/gsf.py]: reverted back.

File: python/graphstorm/model/embed.py [hunk 14]
```diff
@@ -163,6 +164,75 @@ def use_wholegraph_sparse_emb(self):
...
+    def prepare(self, _):
+        """ Preparing input layer for training or inference.
+
+        The input layer can pre-compute edge features in the preparing step
+        if needed. For example pre-compute all BERT embeddings
+
+        Default action: Do nothing
+        """
+
+    def get_general_dense_parameters(self):
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/embed.py:191]: Tested with no edge feature name setting, it will return an empty list. No DDP error found.

File: tests/unit-tests/data_utils.py [hunk 15]
```diff
@@ -491,6 +528,24 @@ def partion_and_load_distributed_graph(hetero_graph, dirname, graph_name='dummy'
     dist_graph = dist.DistGraph(graph_name=graph_name, part_config=part_config)
     return dist_graph, part_config
 
+def generate_special_dummy_dist_graph(dirname, graph_name='special_dummy'):
```
Comment (classicsong) [inline tests/unit-tests/data_utils.py:531]: ditto

File: tests/unit-tests/test_dataloading.py [hunk 16]
```diff
@@ -490,9 +493,15 @@ def test_GSgnnData_edge_feat():
         for _, _, blocks in dataloader:
             edge_feat_list = gdata.get_blocks_edge_feats(blocks, efeat_fields)
             assert len(edge_feat_list) == 2
-            assert edge_feat_list[0][('n0', 'r0', 'n1')].shape[1] == 2
+            if ('n0', 'r0', 'n1') in edge_feat_list[0]:
```
Comment (classicsong) [inline tests/unit-tests/test_dataloading.py:496]: Why there is an if here?

File: tests/unit-tests/test_dataloading.py [hunk 17]
```diff
@@ -532,9 +544,62 @@ def test_GSgnnData_edge_feat():
...
+        dataloader = GSgnnNodeDataLoader(gdata, target_idx, [100, 100], 10,
+                                        label_field='label',
+                                        train_task=False)
+        efeat_fields = {
+            ('n0', 'r0', 'n1'): ['feat'],
+            ('n0', 'r1', 'n1'): ['feat']
+        }
+        for _, _, blocks in dataloader:
+            edge_feat_list = gdata.get_blocks_edge_feats(blocks, efeat_fields)
+            assert len(edge_feat_list) == 2
```
Comment (classicsong) [inline tests/unit-tests/test_dataloading.py:558]: You should also check edge_feat_list[0]

File: tests/unit-tests/test_dataloading.py [hunk 18]
```diff
@@ -532,9 +544,62 @@ def test_GSgnnData_edge_feat():
...
+            assert edge_feat_list[1] == {}
+
     # after test pass, destroy all process group
     th.distributed.destroy_process_group()
 
+def test_GSgnnData_edge_feat2():
+    """ Test GSgnnData built-in functions.
+
```
Comment (classicsong) [inline tests/unit-tests/test_dataloading.py:566]: Can you be more specific about the difference between test_GSgnnData_edge_feat and test_GSgnnData_edge_feat2?

File: tests/unit-tests/test_dataloading.py [hunk 19]
```diff
@@ -532,9 +544,62 @@ def test_GSgnnData_edge_feat():
...
+                                         10,
+                                        label_field='label',
+                                        train_task=False)
+        efeat_fields = {
+            ('n1', 'r1', 'n2'): ['feat'],
+        }
+        for _, _, blocks in dataloader:
+            edge_feat_list = gdata.get_blocks_edge_feats(blocks, efeat_fields)
+            assert len(edge_feat_list) == 2
+            assert edge_feat_list[0] == {}
```
Comment (classicsong) [inline tests/unit-tests/test_dataloading.py:597]: You should also test edge_feat_list[1]

File: python/graphstorm/model/embed.py [hunk 20]
```diff
@@ -618,14 +614,10 @@ def __init__(self,
 
         # set projection weights on edge features
         for canonical_etype in g.canonical_etypes:
-            feat_dim = 0
             if self.feat_size[canonical_etype] > 1:
                 feat_dim = self.feat_size[canonical_etype]
-            if feat_dim > 0:
                 input_projs = nn.Parameter(th.Tensor(feat_dim, self.embed_size))
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/model/embed.py:619]: I will use xavier_uniform to initialize them.

File: tests/unit-tests/test_dataloading.py (hunk 16 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_dataloading.py:496]: Because the test dummy data are randomly generated, and blocks are randomly sampled, the sampled subgraph cannot make sure features always existing for some edge types. Therefore, need to check if the edge type exist first.

File: tests/unit-tests/test_dataloading.py (hunk 17 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_dataloading.py:558]: added checking.

File: tests/unit-tests/test_dataloading.py (hunk 18 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_dataloading.py:566]: add more descriptions.

File: tests/unit-tests/test_dataloading.py (hunk 19 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_dataloading.py:597]: added checking.

**Review by classicsong — APPROVED** [review]
